### ISSUE TITLE
StackOverflowError in KdTree running InteriorPointAreaPerfTest

### ISSUE THREAD

**Issue by FObermaier** [locationtech/jts#602]
Running the `InteriorPointAreaPerfTest` leads to the following output:
```
Interior Point Area perf test
SineStar: origin: (100.0, 100.0)  size: 100.0  # arms: 20  arm ratio: 0.3
# Iterations: 100

Running with # pts 11
runTest1 : 59 ms

Running with # pts 101
runTest1 : 3 ms

Running with # pts 1001
runTest1 : 5 ms

Running with # pts 10001
runTest1 : 9 ms

Running with # pts 100001
runTest1 : 65 ms
Exception in thread "main" java.lang.StackOverflowError
	at org.locationtech.jts.index.kdtree.KdTree.queryNodePoint(KdTree.java:330)
	at org.locationtech.jts.index.kdtree.KdTree.queryNodePoint(KdTree.java:327)
	... many, many more
	at org.locationtech.jts.index.kdtree.KdTree.queryNodePoint(KdTree.java:330)
	at org.locationtech.jts.index.kdtree.KdTree.queryNodePoint(KdTree.java:327)
```

~I'm afraid there are other tests in the performance section that fail, too.~
Double checked, it is just this test.

**Comment by dr-jts** [comment]
The performance tests aren't for automatic validation - they are just useful for manual testing of performance issues and comparisons.

That said, this is due to the `GeometryPrecisionReducer` being changed to use OverlayNG to do precision reduction on polygons.  This uses the `SnapRoundingNoder`, which in turn uses a KdTree.  It is interesting that querying the `KdTree` is so stack-intensive.  This indicates that perhaps it would be better to query the tree using an iterative approach rather than a recursive call.  It might also indicate that the KD-tree is not ideal for this use.

And perhaps another thing to ponder is whether it would be better to try a pointwise precision reduction first, and only snap-round if the result is invalid.  That's a tough call - it's a costly way to see if snap-rounding is needed.

It also indicates that using a precision reduction to generate test data for this test isn't ideal (since it takes far longer than the test itself!)

Are the other performance test failures similar?

**Comment by dr-jts** [comment]
After some investigation, I think this is due to the `KdTree` being built in a very unbalanced way, thus causing a large depth of recursion.  But it's odd, because running a similar case in the TestBuilder doesn't behave this way - it has a much better balanced tree.  More investigation required.

The good news is that I haven't seen this problem occur in other testing (which has processed some pretty big geometry).

**Comment by dr-jts** [comment]
Further investigation shows that KdTree is too unbalanced to provide good performance for very large geometries.  The extreme depth of the tree is what causes the `StackOverflowError`.

It is possible to use a `Quadtree` instead, at the cost of a bit of performance. Another alternative is use `KdTree` for small geometries, and `Quadtree` for large ones.  Alternatively, a balanced KD-tree approach could be used, or perhaps a point-specific packed R-tree.  Both of those would require development, however.  And they would be require another structure  ( a map) to be used to determine the initial set of unique `HotPixel`s.

**Comment by dr-jts** [comment]
@FObermaier  do you know what other performance tests were failing?

**Comment by dr-jts** [comment]
> Double checked, it is just this test.

Good to know, thanks.

**Comment by dr-jts** [comment]
Fixed by #605
